### PR TITLE
feat: reason for removal transformations

### PR DESCRIPTION
**Pull request by andrewmorris43**
## Description

Changes in this PR include:
- Adds transformation logic for 4 chained rules.
- Rules 1 and 2 perform multiple transformations when failing. Part of the transformation includes a db lookup for a BSO code.
- Rules 3 and 4 trigger an exception to be raised when failing.
- Adds unit tests for all 4 new rules.
- Fixes an existing issue when making multiple transformations.
- Changes all the `transformRules.json` file rules to trigger on failure, rather than on success.
- A few tweaks to do with the new queue functionality to run things locally.

### Notes
- Db lookup logic for transformation rules is due to change, so the implementation here is temporary.
- To successfully test this locally a few other rules need to be tweaked, although I believe some rule changes are required anyway.

## Context
[DTOSS-4771](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-4771)
[DTOSS-4573](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-4573)
[DTOSS-4574](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-4574)
[DTOSS-4576](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-4576)
[DTOSS-4577](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-4577)

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

**Review by Warren-Pitterson — APPROVED** [review]

File: application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/TransformDataService.cs [hunk 1]
```diff
@@ -94,26 +95,49 @@ public async Task<CohortDistributionParticipant> TransformParticipantAsync(Cohor
...
+
+        if (!string.IsNullOrEmpty(exceptionMessage))
+        {
+            var participantCsvRecord = new ParticipantCsvRecord
+            {
+                Participant = new Participant(participant),
+                FileName = "",
+            };
+
+            try
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/TransformDataService.cs:123]: I don't think this try catch block makes much sense as you are calling another method that would also fail if we fail to log the error to the table?

Comment (andrewmorris43) [reply application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/TransformDataService.cs:123]: I'll remove it, cheers

**Review by will-larkin-nhs — CHANGES_REQUESTED** [review]
Please reverse the change to make the rules trigger on failure instead of success. It has been unanimously decided in the retros that this is the wrong way to do the rules

File: application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/transformRules.json [hunk 2]
```diff
@@ -433,6 +447,105 @@
             }
           }
         }
+      },
+      {
+        "RuleName": "1.ReasonForRemoval.NonFatal",
```
Comment (will-larkin-nhs) [inline application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/transformRules.json:452]: Please change the names to something more descriptive

**Comment by andrewmorris43** [comment]
> Please reverse the change to make the rules trigger on failure instead of success. It has been unanimously decided in the retros that this is the wrong way to do the rules

@will-larkin-nhs I must have missed that retro where that was decided. I would have argued that it makes more sense to have the rules trigger on failure in this scenario. It makes it much easier to just use the built in ExceptionMessage from rules engine to trigger the validation exception log we need.

I can change it back to trigger on success, but that will involve reverting back to my hacky way to trigger an exception log - not ideal! Plus it doesn't seem right to me to trigger an exception on success.

File: application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/transformRules.json (hunk 2 above)
Comment (andrewmorris43) [reply application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/transformRules.json:452]: Will have a think for a better name

**Comment by andrewmorris43** [comment]
@will-larkin-nhs Ok cheers, I'll revert things.

**Review by will-larkin-nhs — DISMISSED** [review]

**Review by will-larkin-nhs — APPROVED** [review]

**Review by Warren-Pitterson — DISMISSED** [review]

File: application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/TransformDataService.cs [hunk 3]
```diff
@@ -88,30 +88,45 @@ public async Task<CohortDistributionParticipant> TransformParticipantAsync(Cohor
...
+
+        if (result is Exception exception)
+        {
+            var participantCsvRecord = new ParticipantCsvRecord
+            {
+                Participant = new Participant(participant),
+                FileName = "",
+            };
+
+            _logger.LogInformation("A transformation rule raised an exception: {ExceptionMessage}", exception.Message);
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CohortDistributionServices/TransformDataService/TransformDataService.cs:120]: should this be LogError if it's raised an Exception? and can we pass the exception in as well please

**Review by Warren-Pitterson — APPROVED** [review]

**Review by MWClayson-NHS — APPROVED** [review]

**Pull request merged**